File: youtube_m3u_generator.py
#!/usr/bin/env python3
"""
YouTube M3U Generator - Android App Logic
Android uygulamasının mantığıyla YouTube sayfalarına erişim
"""

import re
import time
import logging
import requests
from urllib.parse import unquote
import os

# Logging configuration
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(),
        logging.FileHandler('android_style_m3u.log', encoding='utf-8')
    ]
)

class AndroidStyleYouTubeCrawler:
    def __init__(self):
        self.links_file = "links.txt"
        self.output_file = "youtube_streams.m3u"
        
    def read_channels(self):
        """links.txt dosyasını oku"""
        channels = []
        try:
            with open(self.links_file, 'r', encoding='utf-8') as f:
                content = f.read().strip()
                
            channel_blocks = content.split('\n\n')
            
            for block in channel_blocks:
                block = block.strip()
                if not block:
                    continue
                    
                channel_data = {}
                lines = block.split('\n')
                
                for line in lines:
                    line = line.strip()
                    if line.startswith('isim='):
                        channel_data['name'] = line.replace('isim=', '').strip()
                    elif line.startswith('içerik='):
                        channel_data['url'] = line.replace('içerik=', '').strip()
                    elif line.startswith('logo='):
                        channel_data['logo'] = line.replace('logo=', '').strip()
                
                if 'name' in channel_data and 'url' in channel_data:
                    if 'logo' not in channel_data:
                        channel_data['logo'] = ''
                    channels.append(channel_data)
            
            logging.info(f"✅ {len(channels)} kanal bulundu")
            return channels
            
        except Exception as e:
            logging.error(f"❌ links.txt okuma hatası: {str(e)}")
            return []

    def fetch_web_content(self, url):
        """Android uygulamasındaki gibi web içeriğini çek"""
        try:
            logging.info(f"   📡 Sayfa çekiliyor: {url}")
            
            # Android tarayıcı benzeri headers
            headers = {
                'User-Agent': 'Mozilla/5.0 (Linux; Android 10; SM-G973F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36',
                'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
                'Accept-Language': 'tr-TR,tr;q=0.9,en-US;q=0.8,en;q=0.7',
                'Accept-Encoding': 'gzip, deflate, br',
                'Connection': 'keep-alive',
                'Upgrade-Insecure-Requests': '1',
                'Sec-Fetch-Dest': 'document',
                'Sec-Fetch-Mode': 'navigate',
                'Sec-Fetch-Site': 'none',
                'Cache-Control': 'max-age=0'
            }
            
            response = requests.get(url, headers=headers, timeout=30)
            response.raise_for_status()
            
            logging.info("   ✅ Sayfa başarıyla çekildi")
            return response.text
            
        except Exception as e:
            logging.error(f"   ❌ Sayfa çekme hatası: {str(e)}")
            return None

    def extract_all_urls(self, html_content):
        """HTML içindeki tüm URL'leri çıkar (Android'deki gibi)"""
        try:
            # Tüm URL'leri bulmak için regex pattern
            url_pattern = r'https?://[^\s<>"]+|www\.[^\s<>"]+'
            urls = re.findall(url_pattern, html_content)
            
            logging.info(f"   🔍 {len(urls)} URL bulundu")
            return urls
            
        except Exception as e:
            logging.error(f"   ❌ URL çıkarma hatası: {str(e)}")
            return []

    def extract_hls_urls(self, html_content):
        """HTML içinden HLS URL'lerini çıkar"""
        try:
            # Backslash'leri temizle
            clean_html = html_content.replace('\\', '')
            
            hls_urls = []
            
            # Pattern 1: hlsManifestUrl
            pattern1 = r'"hlsManifestUrl":"(https:[^"]+?m3u8[^"]*?)"'
            matches1 = re.findall(pattern1, clean_html)
            hls_urls.extend(matches1)
            
            # Pattern 2: URL içinde m3u8
            pattern2 = r'"url":"(https:[^"]*?m3u8[^"]*?)"'
            matches2 = re.findall(pattern2, clean_html)
            hls_urls.extend(matches2)
            
            # Pattern 3: Doğrudan m3u8 URL'leri
            pattern3 = r'https://[^"\'\s<>]*?googlevideo\.com[^"\'\s<>]*?m3u8[^"\'\s<>]*'
            matches3 = re.findall(pattern3, clean_html)
            hls_urls.extend(matches3)
            
            # Pattern 4: ytInitialPlayerResponse
            pattern4 = r'ytInitialPlayerResponse\s*=\s*({.+?})\s*;'
            matches4 = re.findall(pattern4, clean_html, re.DOTALL)
            for match in matches4:
                hls_from_json = self.extract_hls_from_json(match)
                if hls_from_json:
                    hls_urls.append(hls_from_json)
            
            # Pattern 5: streamingData
            pattern5 = r'"streamingData":\s*({[^}]+"hlsManifestUrl"[^}]+})'
            matches5 = re.findall(pattern5, clean_html)
            for match in matches5:
                if '"hlsManifestUrl":"' in match:
                    url_match = re.search(r'"hlsManifestUrl":"(https:[^"]+?m3u8[^"]*?)"', match)
                    if url_match:
                        hls_urls.append(url_match.group(1))
            
            # Benzersiz URL'leri döndür
            unique_urls = list(set(hls_urls))
            logging.info(f"   🎯 {len(unique_urls)} HLS URL bulundu")
            
            return unique_urls
            
        except Exception as e:
            logging.error(f"   ❌ HLS URL çıkarma hatası: {str(e)}")
            return []

    def extract_hls_from_json(self, json_str):
        """JSON verisinden HLS URL'sini çıkar"""
        try:
            import json
            # JSON'u temizle
            json_str = re.sub(r'/\*.*?\*/', '', json_str)
            data = json.loads(json_str)
            
            # streamingData içinde ara
            if 'streamingData' in data:
                streaming_data = data['streamingData']
                
                if 'hlsManifestUrl' in streaming_data:
                    hls_url = streaming_data['hlsManifestUrl']
                    if '.m3u8' in hls_url:
                        return hls_url.replace('\\u0026', '&')
                
                # adaptiveFormats içinde ara
                if 'adaptiveFormats' in streaming_data:
                    for fmt in streaming_data['adaptiveFormats']:
                        if 'url' in fmt and '.m3u8' in fmt['url'] and 'googlevideo.com' in fmt['url']:
                            return fmt['url']
            
            return None
            
        except Exception as e:
            logging.error(f"   ❌ JSON parse hatası: {str(e)}")
            return None

    def find_best_hls_url(self, urls):
        """En iyi HLS URL'sini seç"""
        if not urls:
            return None
            
        # Öncelikle en uzun URL'yi seç (genellikle daha fazla parametre = daha kaliteli)
        best_url = max(urls, key=len)
        
        # URL'yi temizle
        best_url = best_url.replace('\\u0026', '&').replace('\\/', '/')
        
        return best_url

    def get_hls_url_android_style(self, url, channel_name):
        """Android uygulaması mantığıyla HLS URL'sini al"""
        logging.info(f"   🌐 Kanal: {channel_name}")
        logging.info(f"   🔗 URL: {url}")
        
        # Desktop URL'ye çevir
        desktop_url = url.replace('//m.youtube.com/', '//www.youtube.com/')
        desktop_url = desktop_url.replace('//youtube.com/', '//www.youtube.com/')
        
        # Web içeriğini çek
        html_content = self.fetch_web_content(desktop_url)
        
        if not html_content:
            return None
        
        # Debug için kaydet
        with open(f"debug_{channel_name.replace(' ', '_')}.html", "w", encoding="utf-8") as f:
            f.write(html_content)
        logging.info(f"   💾 Sayfa kaynağı debug_{channel_name.replace(' ', '_')}.html'ye kaydedildi")
        
        # Tüm URL'leri çıkar
        all_urls = self.extract_all_urls(html_content)
        
        # HLS URL'lerini çıkar
        hls_urls = self.extract_hls_urls(html_content)
        
        if hls_urls:
            # En iyi HLS URL'sini seç
            best_hls_url = self.find_best_hls_url(hls_urls)
            if best_hls_url:
                logging.info(f"   ✅ HLS URL bulundu: {best_hls_url[:80]}...")
                return best_hls_url
        
        logging.info("   ❌ HLS URL bulunamadı")
        return None

    def create_m3u_header(self):
        """M3U dosyası header'ını oluştur"""
        return f"""#EXTM3U
# Title: YouTube Live Streams
# Description: Android uygulama mantığıyla oluşturulmuş YouTube canlı yayın listesi
# Generated: {time.strftime('%Y-%m-%d %H:%M:%S')}
# Method: Android Style Web Crawler

"""

    def write_m3u_file(self, streams):
        """M3U dosyasını yaz"""
        try:
            with open(self.output_file, 'w', encoding='utf-8') as f:
                f.write(self.create_m3u_header())
                
                successful_streams = [s for s in streams if s['hls_url']]
                
                for stream in successful_streams:
                    if stream.get('logo'):
                        f.write(f"#EXTINF:-1 tvg-id=\"{stream['name']}\" tvg-name=\"{stream['name']}\" tvg-logo=\"{stream['logo']}\" group-title=\"YouTube\",{stream['name']}\n")
                    else:
                        f.write(f"#EXTINF:-1 tvg-id=\"{stream['name']}\" tvg-name=\"{stream['name']}\" group-title=\"YouTube\",{stream['name']}\n")
                    f.write(f"{stream['hls_url']}\n\n")
            
            logging.info(f"✅ M3U dosyası oluşturuldu: {self.output_file} ({len(successful_streams)} kanal)")
            return True
            
        except Exception as e:
            logging.error(f"❌ M3U dosyası yazma hatası: {str(e)}")
            return False

    def run(self):
        """Ana çalıştırma fonksiyonu"""
        print("=" * 60)
        print("🚀 YOUTUBE M3U GENERATOR (ANDROID STYLE) - BAŞLIYOR")
        print("=" * 60)
        
        try:
            # Kanal listesini oku
            channels = self.read_channels()
            if not channels:
                logging.error("❌ Hiç kanal bulunamadı!")
                return False

            print("=" * 60)
            print("📡 HLS URL'LERİ ALINIYOR (ANDROID STYLE)...")
            print("=" * 60)

            streams = []
            success_count = 0

            for i, channel in enumerate(channels, 1):
                print(f"\n🎬 KANAL {i}/{len(channels)}: {channel['name']}")
                print(f"   🔗 URL: {channel['url']}")
                if channel.get('logo'):
                    print(f"   🖼️ LOGO: {channel['logo'][:50]}...")
                
                hls_url = self.get_hls_url_android_style(channel['url'], channel['name'])
                
                if hls_url:
                    streams.append({
                        'name': channel['name'],
                        'url': channel['url'],
                        'logo': channel.get('logo', ''),
                        'hls_url': hls_url
                    })
                    success_count += 1
                    print(f"   ✅ BAŞARILI - HLS URL alındı")
                else:
                    streams.append({
                        'name': channel['name'],
                        'url': channel['url'],
                        'logo': channel.get('logo', ''),
                        'hls_url': None
                    })
                    print(f"   ❌ BAŞARISIZ - HLS URL alınamadı")
                
                # Kısa bekleme
                if i < len(channels):
                    time.sleep(2)

            # M3U dosyasını oluştur
            if streams:
                self.write_m3u_file(streams)
                
                # İstatistikleri göster
                print("\n" + "=" * 60)
                print("📊 İSTATİSTİKLER")
                print("=" * 60)
                print(f"📺 Toplam Kanal: {len(channels)}")
                print(f"✅ Başarılı: {success_count}")
                print(f"❌ Başarısız: {len(channels) - success_count}")
                print(f"📈 Başarı Oranı: {(success_count/len(channels))*100:.1f}%")
                print(f"💾 Çıktı Dosyası: {self.output_file}")
                
            return success_count > 0

        except Exception as e:
            logging.error(f"❌ Beklenmeyen hata: {str(e)}")
            return False

def main():
    """Ana fonksiyon"""
    crawler = AndroidStyleYouTubeCrawler()
    success = crawler.run()
    
    if success:
        print("\n🎉 M3U dosyası başarıyla oluşturuldu!")
    else:
        print("\n💥 M3U dosyası oluşturulamadı!")
        exit(1)

if __name__ == "__main__":
    main()

File: youtube_m3u_generator.php
<?php
/**
 * YouTube M3U Generator - PHP cURL Solution
 * No API, No Proxy, No Dependencies
 */

class YouTubeM3UGenerator {
    private $links_file = "links.txt";
    private $output_file = "youtube_streams.m3u";
    
    public function __construct() {
        $this->setupLogging();
    }
    
    private function setupLogging() {
        ini_set('log_errors', 1);
        ini_set('error_log', 'php_m3u_generator.log');
    }
    
    private function log($message, $level = 'INFO') {
        $timestamp = date('Y-m-d H:i:s');
        $logMessage = "$timestamp - $level - $message\n";
        echo $logMessage;
        error_log($logMessage);
    }
    
    public function readChannels() {
        $channels = [];
        try {
            if (!file_exists($this->links_file)) {
                $this->log("❌ links.txt dosyası bulunamadı", 'ERROR');
                return $channels;
            }
            
            $content = file_get_contents($this->links_file);
            $content = trim($content);
            
            $channel_blocks = explode("\n\n", $content);
            
            foreach ($channel_blocks as $block) {
                $block = trim($block);
                if (empty($block)) continue;
                
                $channel_data = [];
                $lines = explode("\n", $block);
                
                foreach ($lines as $line) {
                    $line = trim($line);
                    if (strpos($line, 'isim=') === 0) {
                        $channel_data['name'] = trim(str_replace('isim=', '', $line));
                    } elseif (strpos($line, 'içerik=') === 0) {
                        $channel_data['url'] = trim(str_replace('içerik=', '', $line));
                    } elseif (strpos($line, 'logo=') === 0) {
                        $channel_data['logo'] = trim(str_replace('logo=', '', $line));
                    }
                }
                
                if (isset($channel_data['name']) && isset($channel_data['url'])) {
                    if (!isset($channel_data['logo'])) {
                        $channel_data['logo'] = '';
                    }
                    $channels[] = $channel_data;
                }
            }
            
            $this->log("✅ " . count($channels) . " kanal bulundu");
            return $channels;
            
        } catch (Exception $e) {
            $this->log("❌ links.txt okuma hatası: " . $e->getMessage(), 'ERROR');
            return [];
        }
    }
    
    private function getHLSURL($url, $channel_name) {
        $this->log("   🌐 Sayfa açılıyor: $url");
        
        // Desktop URL'ye çevir
        $desktop_url = str_replace('//m.youtube.com/', '//www.youtube.com/', $url);
        $desktop_url = str_replace('//youtube.com/', '//www.youtube.com/', $desktop_url);
        
        // YouTube'dan sayfayı çek
        $html = $this->fetchYouTubePage($desktop_url);
        
        if (!$html) {
            $this->log("   ❌ Sayfa çekilemedi");
            return null;
        }
        
        // Debug için kaydet
        file_put_contents('debug_page.html', $html);
        $this->log("   📄 Sayfa kaynağı debug_page.html'ye kaydedildi");
        
        // HLS URL'sini ara
        $hls_url = $this->extractHLSFromHTML($html);
        
        if ($hls_url) {
            $this->log("   ✅ HLS URL bulundu: " . substr($hls_url, 0, 80) . "...");
            return $hls_url;
        }
        
        $this->log("   ❌ HLS URL bulunamadı");
        return null;
    }
    
    private function fetchYouTubePage($url) {
        $ch = curl_init();
        
        curl_setopt_array($ch, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_USERAGENT => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            CURLOPT_SSL_VERIFYPEER => false,
            CURLOPT_SSL_VERIFYHOST => false,
            CURLOPT_TIMEOUT => 30,
            CURLOPT_CONNECTTIMEOUT => 15,
            CURLOPT_HTTPHEADER => [
                'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
                'Accept-Language: tr-TR,tr;q=0.9,en;q=0.8',
                'Cache-Control: no-cache',
                'Connection: keep-alive',
                'Upgrade-Insecure-Requests: 1',
            ]
        ]);
        
        $response = curl_exec($ch);
        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $error = curl_error($ch);
        curl_close($ch);
        
        if ($http_code !== 200 || $error) {
            $this->log("   ⚠️ HTTP $http_code - cURL error: $error", 'WARNING');
            return null;
        }
        
        return $response;
    }
    
    private function extractHLSFromHTML($html) {
        // Pattern 1: hlsManifestUrl direkt arama
        if (preg_match('/"hlsManifestUrl":"(https:[^"]+?m3u8[^"]*?)"/', $html, $matches)) {
            $hls_url = str_replace(['\\u0026', '\\/'], ['&', '/'], $matches[1]);
            if (strpos($hls_url, 'googlevideo.com') !== false) {
                return $hls_url;
            }
        }
        
        // Pattern 2: URL içinde m3u8 geçen
        if (preg_match('/"url":"(https:[^"]*?m3u8[^"]*?)"/', $html, $matches)) {
            $hls_url = str_replace(['\\u0026', '\\/'], ['&', '/'], $matches[1]);
            if (strpos($hls_url, 'googlevideo.com') !== false) {
                return $hls_url;
            }
        }
        
        // Pattern 3: ytInitialPlayerResponse içinde arama
        if (preg_match('/ytInitialPlayerResponse\s*=\s*({.+?})\s*;/s', $html, $matches)) {
            $json_str = $matches[1];
            $hls_url = $this->extractHLSFromJSON($json_str);
            if ($hls_url) {
                return $hls_url;
            }
        }
        
        // Pattern 4: window["ytInitialPlayerResponse"] içinde arama
        if (preg_match('/window\["ytInitialPlayerResponse"\]\s*=\s*({.+?})\s*;/s', $html, $matches)) {
            $json_str = $matches[1];
            $hls_url = $this->extractHLSFromJSON($json_str);
            if ($hls_url) {
                return $hls_url;
            }
        }
        
        // Pattern 5: Doğrudan m3u8 URL'leri
        if (preg_match('/https:\/\/[^"\'\s<>]*?googlevideo\.com[^"\'\s<>]*?m3u8[^"\'\s<>]*/', $html, $matches)) {
            return $matches[0];
        }
        
        return null;
    }
    
    private function extractHLSFromJSON($json_str) {
        try {
            // JSON'u temizle
            $json_str = preg_replace('/\/\*.*?\*\//', '', $json_str);
            $data = json_decode($json_str, true);
            
            if (!$data) {
                return null;
            }
            
            // streamingData içinde ara
            if (isset($data['streamingData']['hlsManifestUrl'])) {
                $hls_url = $data['streamingData']['hlsManifestUrl'];
                if (strpos($hls_url, '.m3u8') !== false) {
                    return str_replace('\\u0026', '&', $hls_url);
                }
            }
            
            // adaptiveFormats içinde ara
            if (isset($data['streamingData']['adaptiveFormats'])) {
                foreach ($data['streamingData']['adaptiveFormats'] as $format) {
                    if (isset($format['url']) && strpos($format['url'], '.m3u8') !== false && strpos($format['url'], 'googlevideo.com') !== false) {
                        return $format['url'];
                    }
                }
            }
            
            return null;
            
        } catch (Exception $e) {
            $this->log("   ❌ JSON parse hatası: " . $e->getMessage(), 'ERROR');
            return null;
        }
    }
    
    public function createM3UFile($streams) {
        try {
            $header = "#EXTM3U\n";
            $header .= "# Title: YouTube Live Streams\n";
            $header .= "# Description: PHP ile otomatik oluşturulmuş YouTube canlı yayın listesi\n";
            $header .= "# Generated: " . date('Y-m-d H:i:s') . "\n";
            $header .= "# Total Channels: " . count($streams) . "\n\n";
            
            $content = $header;
            $successful_count = 0;
            
            foreach ($streams as $stream) {
                if ($stream['hls_url']) {
                    if (!empty($stream['logo'])) {
                        $content .= "#EXTINF:-1 tvg-id=\"{$stream['name']}\" tvg-name=\"{$stream['name']}\" tvg-logo=\"{$stream['logo']}\" group-title=\"YouTube\",{$stream['name']}\n";
                    } else {
                        $content .= "#EXTINF:-1 tvg-id=\"{$stream['name']}\" tvg-name=\"{$stream['name']}\" group-title=\"YouTube\",{$stream['name']}\n";
                    }
                    $content .= "{$stream['hls_url']}\n\n";
                    $successful_count++;
                }
            }
            
            if (file_put_contents($this->output_file, $content) !== false) {
                $this->log("✅ M3U dosyası oluşturuldu: {$this->output_file} ($successful_count kanal)");
                return true;
            } else {
                $this->log("❌ M3U dosyası yazılamadı", 'ERROR');
                return false;
            }
            
        } catch (Exception $e) {
            $this->log("❌ M3U dosyası oluşturma hatası: " . $e->getMessage(), 'ERROR');
            return false;
        }
    }
    
    public function run() {
        echo "============================================================\n";
        echo "🚀 YOUTUBE M3U GENERATOR (PHP cURL) - BAŞLIYOR\n";
        echo "============================================================\n";
        
        try {
            // Kanal listesini oku
            $channels = $this->readChannels();
            if (empty($channels)) {
                $this->log("❌ Hiç kanal bulunamadı!", 'ERROR');
                return false;
            }
            
            echo "============================================================\n";
            echo "📡 HLS URL'LERİ ALINIYOR (PHP cURL)...\n";
            echo "============================================================\n";
            
            $streams = [];
            $success_count = 0;
            
            foreach ($channels as $i => $channel) {
                $channel_num = $i + 1;
                $total_channels = count($channels);
                
                echo "\n🎬 KANAL $channel_num/$total_channels: {$channel['name']}\n";
                echo "   🔗 URL: {$channel['url']}\n";
                if (!empty($channel['logo'])) {
                    echo "   🖼️ LOGO: " . substr($channel['logo'], 0, 50) . "...\n";
                }
                
                $hls_url = $this->getHLSURL($channel['url'], $channel['name']);
                
                if ($hls_url) {
                    $streams[] = [
                        'name' => $channel['name'],
                        'url' => $channel['url'],
                        'logo' => $channel['logo'],
                        'hls_url' => $hls_url
                    ];
                    $success_count++;
                    echo "   ✅ BAŞARILI - HLS URL alındı\n";
                } else {
                    $streams[] = [
                        'name' => $channel['name'],
                        'url' => $channel['url'],
                        'logo' => $channel['logo'],
                        'hls_url' => null
                    ];
                    echo "   ❌ BAŞARISIZ - HLS URL alınamadı\n";
                }
                
                // Kısa bekleme
                sleep(2);
            }
            
            // M3U dosyasını oluştur
            if (!empty($streams)) {
                $this->createM3UFile($streams);
                
                // İstatistikleri göster
                echo "\n============================================================\n";
                echo "📊 İSTATİSTİKLER\n";
                echo "============================================================\n";
                echo "📺 Toplam Kanal: " . count($channels) . "\n";
                echo "✅ Başarılı: $success_count\n";
                echo "❌ Başarısız: " . (count($channels) - $success_count) . "\n";
                echo "📈 Başarı Oranı: " . number_format(($success_count / count($channels)) * 100, 1) . "%\n";
                echo "💾 Çıktı Dosyası: {$this->output_file}\n";
                
                return $success_count > 0;
            }
            
            return false;
            
        } catch (Exception $e) {
            $this->log("❌ Beklenmeyen hata: " . $e->getMessage(), 'ERROR');
            return false;
        }
    }
}

// Ana çalıştırma
try {
    $generator = new YouTubeM3UGenerator();
    $success = $generator->run();
    
    if ($success) {
        echo "\n🎉 M3U dosyası başarıyla oluşturuldu!\n";
    } else {
        echo "\n💥 M3U dosyası oluşturulamadı!\n";
        exit(1);
    }
} catch (Exception $e) {
    echo "❌ Kritik hata: " . $e->getMessage() . "\n";
    exit(1);
}
?>
